### DISCUSSION TITLE
socket support

### DISCUSSION THREAD

**Discussion by pottendo** [smuehlst/circle-stdlib#31] [General]
Hi,
thanks for the library and the effort of keeping this alive.
I've used circle-stdlib now to do this: https://github.com/pottendo/pottendo-Pi1541

Today I started writing on a socket layer to wrap Circle CSocket. The background would be to port '[mongoose](https://github.com/cesanta/mongoose/tree/master)' to Circle as it features a few nice functions.
Now I stumbled over your 'circe-socket' branch, which seems not to be continued since a while.

What's the status on this project? Any plans to merge it into the master tree?
(Haven't tried this branch, yet - does it make sense?)

thanks for any feedback!
pottendo

**Comment by smuehlst** [comment]
Hi,

I had the intention to implement the socket layer a while ago. After I had started, something more important stalled this work, and I didn't pick it up since then again.

What is currently published on the https://github.com/smuehlst/circle-stdlib/tree/circle-socket branch is some initial infrastructure code without any functionality. I don't remember exactly but I may have more code locally, I can check that tomorrow.

At the moment I won't promise anything, but potentially I will pick this work up again if there is interest.

Best regards
Stephan

**Reply by pottendo** [reply]
Hi,
sure, I'd be interested. Not only for my project above but other other ideas I have in mind - e.g. a later libcurl support on circle-stdlib could enable quite a bunch of apps to be ported to circle then benefiting from the rapid startup-time, etc.
I can support with testing and even contributing code (as my time & programming skills permit).
thanks for quick response, pottendo

PS: I got mongoose compiling already against circle-stdlib with a layer of dummy socket functions - so that would be my first test-case, to enable a web-server with some mongoose funcs (e.g. serve a directory, etc.).

**Reply by smuehlst** [reply]
I see now that there is first some work required to merge the most recent circle-stdlib release into the `circle-socket` branch. After that I will think about the next steps.

**Comment by pottendo** [comment]
Great, thanks! pottendoAm 30.01.2024 21:51 schrieb Stephan Mühlstrasser ***@***.***>:
I see now that there is first some work required to merge the most recent circle-stdlib release into the circle-socket branch. After that I will think about the next steps.

**Reply by smuehlst** [reply]
@pottendo The most important socket functions are now implemented on the circle-socket branch: `socket()`, `bind()`, `listen()`, `accept()` and `connect()`.

It is clear that this set of functions alone is not sufficient for implementing or porting real applications that use the Posix socket API, but some initial tests should be possible.

The "05-smoketest" sample demonstrates what is working so far:

https://github.com/smuehlst/circle-stdlib/blob/9f68ece1877e2e6c9f994849de0e8deb4e325fba/samples/05-smoketest/kernel.cpp#L766-L960

**Reply by pottendo** [reply]
Thanks for the update. 
I'm traveling and can't test now... will do later this week.
Looking forward to see all this new stuff on the shiny new RP5 circle lib... lots of progress lately happening. Thanks for the efforts!
pottendo

**Comment by pottendo** [comment]
hi,
could try your work now and I can confirm that your smoketest works on my setup (RPi4, Ethernet) - at least the first half with listening and connecting via telnet from my Linux host. The second part (addr 10.0.2.100:5001) may work, but I've not figured out how to do it in my setup.

I jumped further to get a mongoose webserver doing something and made some little progress.
What's noticed so far:
- some build tweaking was necessary 
  - `fcnt()` seems to be coming from 2 parts of the library and conflicts then. So I built my own which just does nothing (for now).
  - The follow macros are needed to compile mongoose:
```
#define FD_CLOEXEC	1
#define F_SETFD		2
#define F_GETFL		4
#define F_SETFL		4
#define O_NONBLOCK	  04000
#define TCP_NODELAY 1
#define FD_SETSIZE 128
```
   of course the values may be incorrect and functions behind are missing, but for a first connection, these are probably not needed.

- `setsockopt(...SO_REUSEADDR...)` needs to be implemented, so I just let it succeed, probably OK for a first test
- `bind(...)` barfs as it get's the local IP passed in `sa_in->sin_addr.s_addr` coming from mongoose, so I overruled to pass INADDR_ANY to make the assertion happy in your code
- `select(...)` is now missing (ENOSYS) and there I'm stuck.

Not sure if this is helpful for you, but I at least gives you thumbs up! Looking forward to any feeback!
bye, pottendo

**Reply by smuehlst** [reply]
Hi @pottendo ,

thanks for giving it a first try. I will look into the problems regarding `fcntl()`, it is not clear to me why it can be defined twice.

At the moment it is expected that any serious application that makes use of the full socket API will not work. `select()` is probably the next thing that will be implemented. But that will take a while, I first have to do some work to catch up with the new Circle release.

**Reply by pottendo** [reply]
Great, looking forward to test further.
Once cleaned up, I'll publish my mongoose experiments, so anyone can benefit.

I did some pre-testing on @rsta2's RPi5 branch (before the release)  with my application, which is actually a circle-stdlib app. 'just' updating circle within circle-stdlib did the trick for me.

Bye, pottendo

**Reply by pottendo** [reply]
...forgot that I needed to do one more change - for even compiling this branch of circle-stdlib:
in `libgloss/circle/stat.cpp` I needed to add `#include <fatfs/ff.h>`.
bye, pottendo

**Reply by pottendo** [reply]
ad the fcntl():
```
  LD    kernel8-rpi4.elf
aarch64-none-elf-ld: ../circle-stdlib/install/aarch64-none-circle/lib/libcirclenewlib.a(io.o): in function `fcntl':
io.cpp:(.text+0xc1c): multiple definition of `fcntl'; ../circle-stdlib/install/aarch64-none-circle/lib/libc.a(lib_a-sysfcntl.o):sysfcntl.c:(.text+0x0): first defined here
```

**Comment by smuehlst** [comment]
> ad the fcntl():
> 
> ```
>   LD    kernel8-rpi4.elf
> aarch64-none-elf-ld: ../circle-stdlib/install/aarch64-none-circle/lib/libcirclenewlib.a(io.o): in function `fcntl':
> io.cpp:(.text+0xc1c): multiple definition of `fcntl'; ../circle-stdlib/install/aarch64-none-circle/lib/libc.a(lib_a-sysfcntl.o):sysfcntl.c:(.text+0x0): first defined here
> ```

Ok, I see that this is a bug, thanks for reporting this. I will fix that with the upcoming circle-stdlib release that syncs with the recent Circle release step 46.

**Comment by smuehlst** [comment]
@pottendo Tbe duplicate definition of `fcntl()` is now fixed in [circle-stdlib v16.3](https://github.com/smuehlst/circle-stdlib/releases/tag/v16.3). Note that the restriction remains that the only `fcntl()` cmd implemented is `F_DUPFD`.

**Comment by smuehlst** [comment]
@pottendo I now picked up development of the socket feature again. @rsta2 supported this with an important addition to Circle's socket implementation, so that it was now possible to implement the `select()` and `poll()` functions.

The new development takes place on the `circle-socket-2` branch.

https://github.com/smuehlst/circle-stdlib/tree/circle-socket-2
 
The implementation is still in a very early stage, and lots of functions from the socket API are not yet implemented. But the implementation is functional enough to run a slightly modified version of the Mongoose web server that you mentioned earlier. This is demonstrated in a corresponding sample application:

https://github.com/smuehlst/circle-stdlib/tree/circle-socket-2/samples/06-socket

**Comment by pottendo** [comment]
Hi,
great to see circle improving also on the networking side.
My project moved on without full socket support and I used the (at the time) existing webserver infrastructure: see here https://github.com/pottendo/pottendo-Pi1541

I may give it a try - not to change my application; that's too much change - but in general.
I'll keep you informed - you decide when to close this ticket.

thanks for efforts you're investing - great work, pottendo

**Comment by pottendo** [comment]
hi,
wanted to checkout your latest socket work - I cloned the repo and switched to the branch,
but it fails to build (_configure -r 3_ and _make all build-stdlib-samples_):
```
/work/src/cbm-retro/scratch/circle-stdlib/libs/circle-newlib/libgloss/circle/stat.cpp: In function 'int _stat(const char*, stat*)':
/work/src/cbm-retro/scratch/circle-stdlib/libs/circle-newlib/libgloss/circle/stat.cpp:19:9: error: 'FILINFO' was not declared in this scope
   19 |         FILINFO finfo;
      |         ^~~~~~~
```
when building libgloss.

Further, I'm not sure what to checkout for building the 06-samples.

minimum I think these:
git clone https://github.com/nlohmann/json.git
git clone https://github.com/cesanta/mongoose.git
other than that I've downloaded the pre-build web-stuff - but that's not yet used before I succeed in building...

thx, pottendo

**Reply by smuehlst** [reply]
Hi @pottendo,

I verified that the following steps result in a successful build from the current `circle-socket-2` branch (commit e351e56161e6cc6a570e0483b4f03af881651680):

```
git clone -b circle-socket-2 --recursive https://github.com/smuehlst/circle-stdlib.git circle-socket-2
cd circle-socket-2
make all
make build-stdlib-samples
```
This also builds the `06-socket` sample.

Maybe you didn't clone all the repositories recursively? This is the safest way to get a working configuration.

**Reply by pottendo** [reply]
hi,
I did recursively checkout, but not the the branch but the repo and switched to the branch. Silly me.
...builds now! thanks as for the quick help, pottendo

**Comment by smuehlst** [comment]
I'm closing this discussion here. It won't be long until I will create a new `circle-stdlib` release with experimental socket support on Codeberg. For further discussion of socket support please create an [issue on Codeberg](https://codeberg.org/larchcone/circle-stdlib/issues).
